pkg/containermeta/store.go

```diff
@@ -0,0 +1,282 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package containermeta
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/status/health"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/DataDog/datadog-agent/pkg/util/retry"
+)
+
+const (
+	retryCollectorInterval = 30 * time.Second
+	pullCollectorInterval  = 5 * time.Second
+)
+
+// Store contains metadata for container workloads.
+type Store struct {
+	mu         sync.RWMutex
+	containers map[string]Container
+	kubePods   map[string]KubernetesPod
+	ecsTasks   map[string]ECSTask
+
+	candidates map[string]Collector
+	collectors []Collector
+
+	eventCh chan []Event
+
+	subscribers map[chan []Event]struct{}
+}
+
+// NewStore creates a new container metadata store. Call Run to start it.
+func NewStore() *Store {
+	candidates := make(map[string]Collector)
+	for id, c := range collectorCatalog {
+		candidates[id] = c()
+	}
+
+	return &Store{
+		candidates:  candidates,
+		collectors:  make([]Collector, len(candidates)),
+		containers:  make(map[string]Container),
+		kubePods:    make(map[string]KubernetesPod),
+		ecsTasks:    make(map[string]ECSTask),
+		subscribers: make(map[chan []Event]struct{}),
+	}
+}
+
+// Run starts the container metadata store.
+func (s *Store) Run(ctx context.Context) {
+	retryTicker := time.NewTicker(retryCollectorInterval)
+	pullTicker := time.NewTicker(pullCollectorInterval)
+	health := health.RegisterLiveness("containermeta-store")
+
+	// Dummy ctx and cancel func until the first pull starts
+	pullCtx, pullCancel := context.WithCancel(ctx)
+
+	go func() {
+		for {
+			select {
+			case <-health.C:
+
+			case <-pullTicker.C:
+				// pullCtx will always be expired at this point
+				// if pullTicker has the same duration as
+				// pullCtx, so we cancel just as good practice
+				pullCancel()
+
+				pullCtx, pullCancel = context.WithTimeout(ctx, pullCollectorInterval)
+				s.pull(pullCtx)
+
+			case evs := <-s.eventCh:
+				s.handleEvents(evs)
+
+			case <-retryTicker.C:
+				s.startCandidates(ctx)
+
+				if len(s.candidates) == 0 {
+					retryTicker.Stop()
+				}
+
+			case <-ctx.Done():
+				retryTicker.Stop()
+				pullTicker.Stop()
+
+				err := health.Deregister()
+				if err != nil {
+					log.Warnf("error de-registering health check: %s", err)
+				}
+
+				return
+			}
+		}
+	}()
+}
+
+// Subscribe returns a channel where container metadata events will be streamed
+// as they happen.
+func (s *Store) Subscribe() chan []Event {
```

Comment:
I think we need to add an interface like `RegisterAdmission()` or something to make the tagger update synchronous and definitely get rid of the force pull in tagger.

Reply:
Can you expand on how `RegisterAdmission()` would work? I understand the problem, but I'm not sure I understand the suggestion.